README.md
# xpat
Chess analysis tool written in Python


src/CHANGELOG.md
# Change log

###[unreleased] 2015-12-05
###Added
 - experimental T5 analysis.

###Changed
 - script logic to keep all scores the same sign as the UCI protocol, only
   changing them to the score from White's viewpoint when printing to the PGN
   file.
   
 - logic of get_score, find_match, and main routines to use tuples instead of
   dictionaries for several structures.
   
###[unreleased] 2015-12-03
###Changed
 - basic move-matching capability to separate scores for black and white.

###Added
 - check_game_end routine and extra logic for treating the final position.
   
 - next_move_score variable for checking the score of the move played to cp
   and engine-matching calculations.


###[unreleased] 2015-12-02
###Changed
 - is_book_position function to accepts a node instead of a board.  Reverted
   back to original setting - I can't ses how to send the starting node instead
   of a board to this function for opening book validation.
   
 - hard-coding of mate score to a settable value.
   
 - variable names for all settings to uppercase.
   
###Added
 - basic move-matching capability.  Largely untested and not fully implemented
   yet but it appears to be working.
   
 - evaluate_position function to de-clutter main() loop.
   
 - open_book function to handle "file not found" exceptions and perform basic
   book validation.
   
###Removed
 - running_board variable from main() loop.  Function and function
   calls changed to accomodate this.
 
###[0.0.2] 2015-12-01
### v0.0.2 released - exclusion of testing/ and external/ directories
   successful.


###[unreleased] 2015-12-01
### Added
 - engine and search depth info to end of game.
   
 - Polyglot book functionality.

### Changed
 - Placed pgn output inside a "with" block instead of keeping it open all the
   time.
   
 - Minor code clean-ups.


###[unreleased] 2015-11-25
### Added
 - .gitattributes to prevent the /testing and /external dircectories from being
   added to future releases.

### Changed
 - get_score to limit the return value between -1000 and +1000 
   centipawns.
   
 - the logic of get_score to return an integer value for centipawns.
   instead of a floating-point value for pawns.  Fixed the pv_score value in
   main() to reflect this.
   
 - Sent engine initialisation to separate function.


###[0.0.1] - 2015-11-25
### v0.0.1 released.




testing/t5_analyser.py
#!/usr/bin/python3

#Copyright 2015 Alan Delaney

# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

import chess, chess.pgn, chess.uci

#Engine settings
engine_name = "stockfish6"
pv_max = 5
search_depth = 10
hash_size = 128


#PGN files
pgnin = open("pgn/twicsample.pgn", "r")
pgnout = open("pgn/gamesout.pgn", "a")


def get_score(eng_score, white_to_move):
	
	if (eng_score.mate):
		
		if (white_to_move):
		
			white_score = 10
			
		else:
			
			white_score = -10
			
	else:
	
		if (white_to_move):
			
			white_score = (int(eng_score.cp))/100
			
		else:
			
			white_score = (int(-(eng_score.cp)))/100
			
	return white_score


def main():
	
	#Engine set-up
	engine = chess.uci.popen_engine(engine_name)
	engine.uci()
	engine.setoption({"Hash":hash_size})
	engine.setoption({"MultiPV":pv_max})
	engine_data = chess.uci.InfoHandler()
	engine.info_handlers.append(engine_data)
	
	next_game = chess.pgn.read_game(pgnin)
	
	i = 0
	
	while next_game:
		
		print ("Analysing game ", i+1)
		this_node = next_game.end().parent
		
		while this_node:
			
			#last_node = this_node.end()
			running_board = this_node.board()
			
			#if (running_board.is_checkmate() or (running_board.is_stalemate())):
			#	
			#	pass
			#	
			#else:
			
			engine.position(running_board)
			engine.go(depth=search_depth)
			
			for j in engine_data.info["pv"]:
				
				pvmove = engine_data.info["pv"][j][0]
				pvscore = get_score (engine_data.info["score"][j], running_board.turn)
				this_node.add_variation (pvmove,starting_comment = str(pvscore))
				
			print (".", end = "", flush=True)
			#break
			#Analyse forwards
			#this_node = this_node.variation(0)
			#Analyse backwards
			this_node = this_node.parent
			
		print()
		
		pgnout.write(str(next_game))
		pgnout.write("\n\n")
		
		i += 1
		next_game = chess.pgn.read_game(pgnin)
		
	print (i, " boards printed.")
	
	engine.quit()
	pgnin.close()
	pgnout.close()


if __name__ == "__main__":
	#main(sys.argv[1:])
	main()





testing/auto_adjudicator.py
#!/usr/bin/python3.5

#Copyright 2015 Alan Delaney

# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

import chess
import chess.pgn
import chess.uci

#Engine settings
engine_name = "stockfish6"
#engine_name = "/usr/local/blackrebel/engines/stockfish-6-linux/Linux/stockfish_6_x\64"
search_depth = 30
hash_size = 1024

#PGN files
pgnin = open("pgn/gamesin.pgn", "r")
pgnout = "pgn/gamesout.pgn"
#pgnin = open("pgn/ptests1.pgn", "r")
#pgnout = open("pgn/ptests1out.pgn", "a")
#pgnin = open("pgn/qtest.pgn", "r")
#pgnout = open("pgn/qtestout.pgn", "a")

#dataout = open("pgn/dataout.txt", "a")

#Settings
white_threshold = 70
black_threshold = -50

def init_engine (eng_name, eng_hash=False):
    '''Engine start-up'''
    new_engine = chess.uci.popen_engine(eng_name)
    #new_engine = chess.uci.popen_engine("./stockfish6")
    new_engine.uci()
    
    if eng_hash:
        
        new_engine.setoption({"Hash":eng_hash})
        
    return new_engine


def get_score(eng_score, white_to_move):
    
    if (eng_score.mate):
        
        if (white_to_move):
        
            white_score = 10
            
        else:
            
            white_score = -10
            
    else:
    
        if (white_to_move):
            
            white_score = int(eng_score.cp)
            
        else:
            
            white_score = int(-(eng_score.cp))
            
        if white_score > 1000:
            
            white_score = 1000
            
        if white_score < -1000
            
            white_score = -1000
            
        

        
    return white_score


def main():

    engine = init_engine("engine_name", 1024)
    engine_data = chess.uci.InfoHandler()
    engine.info_handlers.append(engine_data)
    
    next_game = chess.pgn.read_game(pgnin)
    
    i = 0
    
    while next_game:
        
        end_node = next_game.end()
        running_board = end_node.board()
        
        if not ((running_board.is_checkmate()) or 
                (running_board.is_stalemate())):
            
            print ("Analysing game ", i+1)
            engine.position(running_board)
            engine.go(depth=search_depth)
            print ("Analysis complete.")
            print()
            print()
            
            pv_score = get_score(engine_data.info["score"][1], 
                    running_board.turn)
            end_node.parent.add_variation(end_node.move, starting_comment 
                    = (engine.name + " " + str(engine_data.info["depth"]) 
                    + ": " + str(pv_score)))
            end_node.parent.variation(1).add_variation (engine_data.info
                    ["pv"][1][0])
            #end_node.add_variation (, starting_comment = (engine.name +
            
            if ((pv_score >= white_threshold) and (next_game.headers["Result"] 
                    != '1-0')):
                
                next_game.headers["Result"] = "1-0"
                end_node.comment = end_node.comment + " - auto-adjusted to 1-0"
                
            elif ((pv_score <= black_threshold) and (next_game.headers
                    ["Result"] != '0-1')):
                
                next_game.headers["Result"] = "0-1"
                end_node.comment = end_node.comment + " - auto-adjusted to 0-1"
                
            elif ((pv_score < white_threshold) and (pv_score > black_threshold) 
                    and (next_game.headers["Result"] != "1/2-1/2")):
                
                next_game.headers["Result"] = "1/2-1/2"
                end_node.comment = (end_node.comment + 
                        " - auto-adjusted to 1/2-1/2")
                
            else:
                
                end_node.comment = end_node.comment + " - auto-adjudicator "
                        + "detects no result change necessary"
                
        with open (pgn_out, 'a') as f:
            f.write (str(next_game))
            f.write ("\n\n")
        
        
        i += 1
        next_game = chess.pgn.read_game(pgnin)
        
    print (i, " boards printed.")
    
    engine.quit()
    pgnin.close()
    #pgnout.close()


if __name__ == "__main__":
    #main(sys.argv[1:])
    main()





testing/snapshots/t5_analyser.151122.0318.py
#!/usr/bin/python3

#Copyright 2015 xarxziux

# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

import chess, chess.pgn, chess.uci

PVTot = 5

#pgnin = open("20games.pgn", "r")
#pgnout = open("20games_out.pgn", "a")

pgnin = open("testversions/testgame.pgn", "r")
pgnout = open("testversions/testoutput.pgn", "a")

dataout = open("testversions/testdata.txt", "a")

engine = chess.uci.popen_engine("stockfish6")
engine.uci()

engine.setoption({"Hash":512})
engine.setoption({"MultiPV":PVTot})
engine_data = chess.uci.InfoHandler()
engine.info_handlers.append(engine_data)

next_game = chess.pgn.read_game(pgnin)

i = 0

while next_game:
#while i < 1:
	
	print ("Analysing game ", i+1)
	this_node = next_game
	
	while this_node.variations:
		
		#next_node = this_node.variation(0)
		
		running_board = this_node.board()
		
		if (running_board.is_checkmate()):
			
			if (running_board.turn):
				
				this_node.comment = "White is in checkmate."
				
			else:
				
				this_node.comment = "Black is in checkmate."
				
			break
			
		elif (running_board.is_stalemate()):
			
			this_node.comment = "Position is stalemate."
			break
			
		#else:
		
		#next_game.add_variation (chess.Move.from_uci("0000"), starting_comment = "No comment!")
		#next_node.comment = "No comment!"
		
		engine.position(running_board)
		engine.go(depth=10)
		
		i = 0
		
		for i in engine_data.info["pv"]:
			
			pvmove = engine_data.info["pv"][i][0]
			pvscore = engine_data.info["score"][i].cp
			
			if not running_board.turn:
				
				pvscore = -pvscore
				
			this_node.add_variation (pvmove,starting_comment = str(pvscore))
			
		#print
		#print ("Result = ", next_game.headers["Result"])
		#dataout.write (str(engine_data.info))
		
		# for i in range(1,PVTot+1):
		
		# if i in engine_data.info["pv"]:
		# #print engine_data.info["pv"][i]
		# #print ("PV %i: %s, score: %i") % (i, engine_data.info["pv"][i][0], engine_data.info["score"][i].cp) 
		# print ("PV", i,":",engine_data.info["pv"][i][0], "score:", engine_data.info["score"][i].cp) 
		# #print()
		# #print()
		
		#this_node = next_node
		this_node = this_node.variation(0)
		
	pgnout.write(str(next_game))
	pgnout.write("\n\n")
	next_game = chess.pgn.read_game(pgnin)
	
	# if (running_board.turn):
		
		# print (engine_data.info)
		# #print ("Score =", engine_data.info["score"][1].cp)
		
	# else:
		
		# print ("Score =", engine_data.info)
		# #print ("Score =", -(engine_data.info["score"][1]))
		# #print ("Score =", -(engine_data.info["score"][1].cp))
		
	i += 1
	next_game = chess.pgn.read_game(pgnin)
	
print (i, " boards printed.")

engine.quit()
pgnin.close()
#pgnout.close()
dataout.close()







src/t5_analyser.py
#!/usr/bin/python3

#Copyright 2015 Alan Delaney

# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

#python-chess imports
import chess
import chess.pgn
import chess.uci
import chess.polyglot


#Engine settings
ENGINE_NAME = "stockfish6"
SEARCH_DEPTH = 10
HASH_SIZE = 128
THREAD_COUNT = 1


#Files
INPUT_FILE = "pgn/gamesin.pgn"
OUTPUT_FILE = "pgn/gamesout.pgn"
BOOK_FILE = "polyglot/Elo2400.bin"
#BOOK_FILE = "pgn/gamesout.pgn"


#Tx settings
PV_MAX = 5
PV_MIN = 3
CP_POSITION_CUTOFF = 200
CP_PV_CUTOFF = 100


#General settings
MATE_SCORE = 1000
CP_THRESHOLD = 300


def get_score(eng_score):
    ''' Convert the UCI score output to an integer value bound by +/- MATE_SCORE '''
    
    if (eng_score.mate):
        
        return -MATE_SCORE
            
    else:
        
        return_score = (int(eng_score.cp))
            
        if (return_score > MATE_SCORE):
            
            return_score = MATE_SCORE
            
        if (return_score < -MATE_SCORE):
            
            return_score = -MATE_SCORE
            
    return return_score


def init_engine (eng_settings):
    '''Engine set-up'''
    
    new_engine = chess.uci.popen_engine(eng_settings['name'])
    new_engine.uci()
    
    if 'hash' in eng_settings:
        
        new_engine.setoption({"Hash":eng_settings['hash']})
        
    if 'threads' in eng_settings:
        
        new_engine.setoption({"Threads":eng_settings['threads']})
        
    if 'multipv' in eng_settings:
        
        new_engine.setoption({"MultiPV":eng_settings['multipv']})
        
    else:
        
        new_engine.setoption({"MultiPV":5})
        
    return new_engine


def is_book_position (p_book, c_board):
    ''' Checks if the given position is in the given book '''
    
    try:
        
        p_book.find (c_board)
        return True
        
    except IndexError:
        
        return False
        


def evaluate_position (node, eng):
    
    eng.position(node.board())
    eng.go(depth=SEARCH_DEPTH)


def check_game_end (node):
    ''' Checks whether the given position is a legal end '''
    ''' If the game is a legal, as opposed to an agreed, end, then no engine
        evaluation is necessary '''
    
    if (node.board().is_stalemate() or
            node.board().is_insufficient_material() or
            node.board().can_claim_draw()
            ):
        
        return (True, 0)
        
    if (node.board().is_checkmate()):
        
        return (True, -(MATE_SCORE))
            
    return (False, 0)


def open_book ():
    
    try:
        
        new_book = chess.polyglot.open_reader (BOOK_FILE)
        
    except FileNotFoundError:
        
        return False
        
    #Validate book
    if is_book_position (new_book, chess.Board()):
        
        return new_book
        
    else:
        
        return False
        


def find_match (pvs, match_list):
    ''' Check if the move played is among the PV choices found by the engine '''
    
    #First ignore positions with less than PV_MIN variations
    if (len(pvs) < (PV_MIN + 1)):
        
        match_list[1] += 1
        return False
        
    #Check if the score of the top choice falls outside the position threshold
    if (abs(pvs[1][1]) >= CP_POSITION_CUTOFF):
        
        match_list[1] += 1
        return False
        
    #Check if the score of PV_MIN falls outside the PV cutoff
    if (pvs[1][1] - pvs[PV_MIN][1] > CP_PV_CUTOFF):
        
        match_list[1] += 1
        return False
        
    match_list[0] += 1
    
    i = 2
    
    for var in pvs[1:]:
        
        if (pvs[0][0] == var[0]):
            
            match_list[i] += 1
            return True
            
        i += 1
    return False


def init_match_list ():
    ''' Sets up the match list '''
    ''' The first element is the number of move counted.
        The second is the number of moves rejected.
        Subsequent items show the matches for T1, T2, T3, etc. '''
    
    match_list = [0,0]
    
    i = 0
    
    while (i < PV_MAX):
        
        match_list.append (0)
        i += 1
        
    return match_list


def main():
    
    #Engine set-up
    user_eng_settings = {'name' : ENGINE_NAME,
                         'hash' : HASH_SIZE,
                         'multipv' : PV_MAX,
                         'threads' : THREAD_COUNT
                        }
    engine = init_engine (user_eng_settings)
    engine_data = chess.uci.InfoHandler()
    engine.info_handlers.append(engine_data)
    
    white_match = init_match_list()
    black_match = init_match_list()
    
    #Book set-up
    polyglot_book = open_book ()
    if not polyglot_book:
        
        #open_book returned False
        print ("Bad file name or format for BOOK_FILE.  Exiting.")
        return
    
    pgn_in = open(INPUT_FILE, "r")
    next_game = chess.pgn.read_game(pgn_in)
    
    i = 0
    
    while next_game:
        
        #Skip main() loop for test purposes
        #break
        
        print ("Analysing game ", i+1)
        this_node = next_game.end()
        pv_list = []
        
        if is_book_position (polyglot_book, this_node.board()):
            
            this_node.comment = "Book position"
            break
            
        legal_end = check_game_end(this_node)
        
        if legal_end[0]:
            
            pv_list.append ((this_node.move, -(legal_end[1])))
            
        else:
            
            evaluate_position (this_node, engine)
            pv_list.append ((this_node.move, -(get_score (engine_data.info["score"][1]))))
            #print (engine_data.info["score"])
            #end_score = -(get_score (engine_data.info["score"][1]))
            #pv_list.append ((this_node.move, end_score))
            
        this_node = this_node.parent
        
        while this_node:
            
            #This line checks if the next move results in a book position
            if is_book_position (polyglot_book, this_node.variation(0).board()):
                
                this_node.variation(0).comment = "Book position"
                break
                
            evaluate_position (this_node, engine)
            
            for next_pv in engine_data.info["pv"]:
                
                pv_move = engine_data.info["pv"][next_pv][0]
                pv_score = (get_score (engine_data.info["score"][next_pv]))
                pv_list.append ((pv_move, pv_score))
                
                output_score = pv_score/100
                
                if not (this_node.board().turn):
                    
                    output_score = -output_score
                    
                this_node.add_variation (pv_move, starting_comment = str(output_score))
                
            if (this_node.board().turn):
                
                #print ("pv_list =", pv_list)
                find_match(pv_list, white_match)
                
            else:
                
                #print ("pv_list =", pv_list)
                find_match(pv_list, black_match)
                
            print (".", end = "", flush=True)
            pv_list = [(this_node.move, -(pv_list[1][1]))]
            this_node = this_node.parent
            
        print()
        print (white_match)
        print (black_match)
        print()
        
        next_game.end().comment = ("T5 analysis performed by '" + engine.name + 
                "' using a search depth of " + str(SEARCH_DEPTH) + ".")
        
        with open (OUTPUT_FILE, 'a') as pgn_out:
            
            pgn_out.write(str(next_game))
            pgn_out.write("\n\n")
        
        i += 1
        next_game = chess.pgn.read_game(pgn_in)
        
    print (i, " boards printed.")
    
    engine.quit()
    pgn_in.close()


if __name__ == "__main__":
    #main(sys.argv[1:])
    main()





auto_adjudicator.py
#!/usr/bin/python3

import chess, chess.pgn, chess.uci

pgnin = open("20games.pgn")
#pgnout = open("20games_out.pgn", "a")

engine = chess.uci.popen_engine("stockfish6")
engine.uci()

engine.setoption({"Hash":512})
engine_data = chess.uci.InfoHandler()
engine.info_handlers.append(engine_data)

next_game = chess.pgn.read_game(pgnin)

i = 0



while next_game:

	end_node = next_game.end()
	running_board = end_node.board()
	
	print ("Analysing game ", i+1)
	
	if (running_board.is_checkmate()):
		
		if (running_board.turn):
			
			print ("Black is in checkmate.")
			
		else:
			
			print ("White is in checkmate.")
			
		i += 1
		next_game = chess.pgn.read_game(pgnin)
		continue
		
	elif (running_board.is_stalemate()):
		
		print ("Position is stalemate.")
		print()
		i += 1
		next_game = chess.pgn.read_game(pgnin)
		continue
		
	engine.position(running_board)
	engine.go(depth=10)
	
	print ("Result = ", next_game.headers["Result"])
	
	if (running_board.turn):
		
		print ("Score =", engine_data.info["score"][1].cp)
		
	else:
		
		print ("Score =", -(engine_data.info["score"][1].cp))
		
	print ()
	
	i += 1
	next_game = chess.pgn.read_game(pgnin)
	
print (i, " boards printed.")
	
engine.quit()
pgnin.close()
#pgnout.close()







pgnparser.data.py
#!/usr/bin/python3

import chess, chess.pgn, chess.uci

pgn = open("schemingmind.pgn")

first_game = chess.pgn.read_game(pgn)

#print "first_game:"
#print dir(first_game)
#print (first_game)    #Print PGN list
'''
first_game:
['__class__', '__delattr__', '__dict__', '__doc__', '__format__', '__getattribute__', '__hash__', '__init__', '__module__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', '__weakref__', 'add_main_variation', 'add_variation', 'board', 'board_cached', 'comment', 'demote', 'end', 'errors', 'export', 'has_variation', 'headers', 'is_main_line', 'is_main_variation', 'move', 'nags', 'parent', 'promote', 'promote_to_main', 'remove_variation', 'root', 'san', 'setup', 'starting_comment', 'starts_variation', 'variation', 'variations']
[Event "For New Members (#94)  10d+1d/m"]
[Site "SchemingMind.com"]
[Date "2014.05.09"]
[Round "-"]
[White "hardpawn"]
[Black "xarxziux"]
[Result "0-1"]
[WhiteElo "1640"]
[BlackElo "1938"]
[ECO "B07m"]
[WhiteCountry "AUS"]
[BlackCountry "IRL"]
[WhiteRD "109"]
[BlackRD "119"]
[GameID "358192"]
[Annotator "Critter 1.6a 64-bit (60 sec)"]

1. e4 d6 2. d4 Nf6 3. Nc3 e5 4. dxe5 dxe5 5. Nf3 Qxd1+ 6. Nxd1 Nxe4 7. Nxe5 Nd7 8. Nf3 Bc5 9. Be3 O-O 10. Bd3 Ndf6 11. O-O Bg4 12. Be2 Bxe3 13. Nxe3 Bxf3 14. Bxf3 Nd2 15. Rfd1 Nxf3+ 16. gxf3 Rfd8 17. c4 c6 18. Nf5 Kf8 19. Nd6 b6 20. Rd3 Rd7 21. Rad1 Rad8 22. b4 c5 23. bxc5 bxc5 24. Kf1 Ne8 25. Ke2 Rxd6 26. R1d2 Rxd3 27. Rxd3 Rxd3 28. Kxd3 Ke7 29. Kc3 Nd6 30. Kb3 Ke6 31. Ka4 Nxc4 32. Kb5 Kd5 33. a4 Nd6+ 34. Ka6 c4 0-1
'''

#print "Variation:"
#print first_game.variation
#print "Variations:"
#print first_game.variations
'''
Variation:
<bound method Game.variation of <chess.pgn.Game object at 0x7f15cd090390>>
Variations:
[<chess.pgn.GameNode object at 0x7f15ccfc8690>]
'''

#print "Variation:"
#print dir(first_game.variation)
#print "Variations:"
#print dir(first_game.variations)
'''
Variation:
['__call__', '__class__', '__cmp__', '__delattr__', '__doc__', '__format__', '__func__', '__get__', '__getattribute__', '__hash__', '__init__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__self__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', 'im_class', 'im_func', 'im_self']
Variations:
['__add__', '__class__', '__contains__', '__delattr__', '__delitem__', '__delslice__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__getitem__', '__getslice__', '__gt__', '__hash__', '__iadd__', '__imul__', '__init__', '__iter__', '__le__', '__len__', '__lt__', '__mul__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__reversed__', '__rmul__', '__setattr__', '__setitem__', '__setslice__', '__sizeof__', '__str__', '__subclasshook__', 'append', 'count', 'extend', 'index', 'insert', 'pop', 'remove', 'reverse', 'sort']
'''

#print "Variation:"
#print first_game.variation(0)
#print "Variations:"
#print first_game.variations(0)
'''
d6 2. d4 Nf6 3. Nc3 e5 4. dxe5 dxe5 5. Nf3 Qxd1+ 6. Nxd1 Nxe4 7. Nxe5 Nd7 8. Nf3 Bc5 9. Be3 O-O 10. Bd3 Ndf6 11. O-O Bg4 12. Be2 Bxe3 13. Nxe3 Bxf3 14. Bxf3 Nd2 15. Rfd1 Nxf3+ 16. gxf3 Rfd8 17. c4 c6 18. Nf5 Kf8 19. Nd6 b6 20. Rd3 Rd7 21. Rad1 Rad8 22. b4 c5 23. bxc5 bxc5 24. Kf1 Ne8 25. Ke2 Rxd6 26. R1d2 Rxd3 27. Rxd3 Rxd3 28. Kxd3 Ke7 29. Kc3 Nd6 30. Kb3 Ke6 31. Ka4 Nxc4 32. Kb5 Kd5 33. a4 Nd6+ 34. Ka6 c4
Variations:
Traceback (most recent call last):
  File "pgnparser.data.py", line 58, in <module>
    print first_game.variations(0)
TypeError: 'list' object is not callable)
'''

#print "Variation:"
#print first_game.variation(0)
#print first_game.variation(1)
#print first_game.variation(2)
'''
Variation:
d6 2. d4 Nf6 3. Nc3 e5 4. dxe5 dxe5 5. Nf3 Qxd1+ 6. Nxd1 Nxe4 7. Nxe5 Nd7 8. Nf3 Bc5 9. Be3 O-O 10. Bd3 Ndf6 11. O-O Bg4 12. Be2 Bxe3 13. Nxe3 Bxf3 14. Bxf3 Nd2 15. Rfd1 Nxf3+ 16. gxf3 Rfd8 17. c4 c6 18. Nf5 Kf8 19. Nd6 b6 20. Rd3 Rd7 21. Rad1 Rad8 22. b4 c5 23. bxc5 bxc5 24. Kf1 Ne8 25. Ke2 Rxd6 26. R1d2 Rxd3 27. Rxd3 Rxd3 28. Kxd3 Ke7 29. Kc3 Nd6 30. Kb3 Ke6 31. Ka4 Nxc4 32. Kb5 Kd5 33. a4 Nd6+ 34. Ka6 c4
Traceback (most recent call last):
  File "pgnparser.data.py", line 70, in <module>
    print first_game.variation(1)
  File "/usr/local/lib/python2.7/dist-packages/chess/pgn.py", line 180, in variation
    raise KeyError("variation not found")
KeyError: 'variation not found'
'''

#print "Variation:"
#print first_game.variation(0)(0)
'''
Variation:
Traceback (most recent call last):
  File "pgnparser.data.py", line 84, in <module>
    print first_game.variation(0)(0)
TypeError: 'GameNode' object is not callable
'''

node = first_game
running_board = chess.Board()

#print "node:"
#print node
#print "running_board:"
#print running_board
'''
node:
[Event "For New Members (#94)  10d+1d/m"]
[Site "SchemingMind.com"]
[Date "2014.05.09"]
[Round "-"]
[White "hardpawn"]
[Black "xarxziux"]
[Result "0-1"]
[WhiteElo "1640"]
[BlackElo "1938"]
[ECO "B07m"]
[WhiteCountry "AUS"]
[BlackCountry "IRL"]
[WhiteRD "109"]
[BlackRD "119"]
[GameID "358192"]
[Annotator "Critter 1.6a 64-bit (60 sec)"]

1. e4 d6 2. d4 Nf6 3. Nc3 e5 4. dxe5 dxe5 5. Nf3 Qxd1+ 6. Nxd1 Nxe4 7. Nxe5 Nd7 8. Nf3 Bc5 9. Be3 O-O 10. Bd3 Ndf6 11. O-O Bg4 12. Be2 Bxe3 13. Nxe3 Bxf3 14. Bxf3 Nd2 15. Rfd1 Nxf3+ 16. gxf3 Rfd8 17. c4 c6 18. Nf5 Kf8 19. Nd6 b6 20. Rd3 Rd7 21. Rad1 Rad8 22. b4 c5 23. bxc5 bxc5 24. Kf1 Ne8 25. Ke2 Rxd6 26. R1d2 Rxd3 27. Rxd3 Rxd3 28. Kxd3 Ke7 29. Kc3 Nd6 30. Kb3 Ke6 31. Ka4 Nxc4 32. Kb5 Kd5 33. a4 Nd6+ 34. Ka6 c4 0-1
running_board:
r n b q k b n r
p p p p p p p p
. . . . . . . .
. . . . . . . .
. . . . . . . .
. . . . . . . .
P P P P P P P P
R N B Q K B N R
'''
















#'''
engine = chess.uci.popen_engine("stockfish6")
engine.uci()

PVTot = 5
#engine.setoption({"Hash":512}, {"MultiPV":PVTot})
engine.setoption({"MultiPV":PVTot})
engine.setoption({"Hash":512})

engine_data = chess.uci.InfoHandler()
engine.info_handlers.append(engine_data)
#print dir(engine.infohandlers)

move_count = 1
white_to_move = True


while node.variations:
    next_node = node.variation(0)
    
    if white_to_move:
      print ("White move "),
      print (move_count, ":", node.move),
      #print (": "),
      white_to_move = False
    else:
      print ("Black move"),
      print (move_count, ":", node.move),
      #print (": "),
      white_to_move = True
      move_count = move_count + 1
      
    #print (node.move)
    #print
    running_board = node.board()
    #print(running_board)
    engine.position(running_board)
    engine.go(depth=10)
    #print engine.bestmove
    #print engine_data.info[1]

    for i in range(1,PVTot+1):
         
         if i in engine_data.info["pv"]:
             #print engine_data.info["pv"][i]
             #print ("PV %i: %s, score: %i") % (i, engine_data.info["pv"][i][0], engine_data.info["score"][i].cp) 
             print ("PV", i,":", engine_data.info["pv"][i][0], "score: ", engine_data.info["score"][i].cp) 
             #print 
             #print
    
    #print engine.multipv
    #print engine_data.BestMove
    #print engine_data.pv(3)
    #print (running_board)
    
    print()
    print()
    #game_list = game_list + " " + node.board().san(next_node.move)
    #print(node.board().san(next_node.move))
    #print (game_list)
    node = next_node

#print engine_data.info
#print engine_data.pv(1)
#print engine_data.pv(2)
#print engine_data.pv(3)




#'''
engine.quit()
pgn.close()


